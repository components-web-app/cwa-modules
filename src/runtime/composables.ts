import { useNuxtApp } from '#imports'

export const useCwa = () => useNuxtApp().$cwa
