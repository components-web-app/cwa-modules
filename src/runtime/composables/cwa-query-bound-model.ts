import { useRouter, useRoute, type LocationQueryValue } from 'vue-router'
import { computed, ref, watch } from 'vue'
import { debounce } from 'lodash-es'

type ModelOps = {
  defaultValue?: any
  delay?: number
}

export const useQueryBoundModel = (queryParam: string, ops?: ModelOps) => {
  const route = useRoute()
  const router = useRouter()
  let debounced: any

  const exp = new RegExp(`^${queryParam}\\[([a-zA-Z0-9]+)]$`, 'i')
  const matchingQueryParams = computed(() => {
    if (route.query[queryParam]) {
      return [queryParam]
    }
    const allQueryParams = Object.keys(route.query)
    return allQueryParams.filter(param => exp.test(param))
  })

  const matchedQueryParamValue = computed(() => {
    const matchingParams = matchingQueryParams.value
    if (!matchingParams || !matchingParams.length) {
      return null
    }
    if (route.query[queryParam]) {
      return route.query[queryParam]
    }
    const matches = matchingParams[0].match(exp)
    if (!matches || matches.length < 2) {
      return null
    }
    const matchingObjectKey = matches[1]
    if (matchingParams.length && matchingObjectKey) {
      return { [matchingObjectKey]: route.query[matchingParams[0]] }
    }
    return null
  })

  const model = ref(matchedQueryParamValue.value !== null ? matchedQueryParamValue.value : ops?.defaultValue)

  watch(matchedQueryParamValue, (newValue) => {
    model.value = newValue
  })
  watch(model, async (newValue) => {
    if (debounced) {
      debounced.cancel()
    }
    const filteredKeys = Object.keys(route.query).filter(key => !matchingQueryParams.value.includes(key))
    const newQuery: { [key: string]: LocationQueryValue|LocationQueryValue[] } = {}
    for (const retainedKey of filteredKeys) {
      newQuery[retainedKey] = route.query[retainedKey]
    }

    const isObj = Object.prototype.toString.call(newValue) === '[object Object]'
    if (isObj) {
      for (const newValueKey of Object.keys(newValue)) {
        newQuery[`${queryParam}[${newValueKey}]`] = newValue[newValueKey]
      }
    } else {
      newQuery[queryParam] = newValue
    }

    debounced = debounce(async () => {
      await router.replace({ query: newQuery })
    }, ops?.delay || 10)
    await debounced()
  })

  return {
    model
  }
}
