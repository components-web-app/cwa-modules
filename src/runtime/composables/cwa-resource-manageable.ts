import { getCurrentInstance, onBeforeUnmount, onMounted } from 'vue'
import logger from 'consola'
import ManageableComponent, { ManageableComponentOptions } from '../admin/manageable-component'
import { CwaCurrentResourceInterface } from '../storage/stores/resources/state'
import { useCwa } from './cwa'

/**
 * @internal
 * @description Advanced usage - usually this composable will be initialised from useCwaResource where disableManager does not equal true. Primarily separated for the ComponentGroup component
 */
export const useCwaResourceManageable = (iri?: string, options?: ManageableComponentOptions) => {
  const proxy = getCurrentInstance()?.proxy
  if (!proxy) {
    logger.error('Cannot initialise manager for resource. Instance is not defined')
    return
  }

  const manageableComponent = new ManageableComponent(proxy, useCwa(), options)

  onMounted(() => {
    if (iri) {
      manageableComponent.init(iri)
    }
  })

  onBeforeUnmount(() => {
    manageableComponent.clear()
  })

  return {
    resourceWatchHandler: (resource: CwaCurrentResourceInterface) => {
      const iri = resource?.data?.['@id']
      iri && manageableComponent.init(iri)
    },
    manager: manageableComponent
  }
}
