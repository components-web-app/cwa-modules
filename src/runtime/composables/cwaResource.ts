import { useCwa } from './cwa'

export type IriProp = {
 iri: string
}

interface cwaResourceUtilsOps {}

export const useCwaResourceUtils = (_: cwaResourceUtilsOps) => {
  return {
    // this needs to be a function so useCwa is not called early - would get issues from ComponentPosition and more
    getResource: (id:string) => useCwa().resources.getResource(id)
  }
}
