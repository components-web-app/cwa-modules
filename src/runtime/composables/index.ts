import { useNuxtApp } from '#imports'

export const useCwa = () => useNuxtApp().$cwa
export const useCwaResources = () => useNuxtApp().$cwa.resources
export const useCwaResourcesManager = () => useNuxtApp().$cwa.resourcesManager
export const useCwaAuth = () => useNuxtApp().$cwa.auth
export const useCwaForms = () => useNuxtApp().$cwa.forms
