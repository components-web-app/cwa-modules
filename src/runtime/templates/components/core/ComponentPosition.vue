<template>
  <!--CWA_MANAGER_START_POSITION-->
  <ResourceLoader v-if="componentIri" ref="resourceLoader" :iri="componentIri" :manager="manager" component-prefix="CwaComponent" />
  <ComponentPlaceholder v-else-if="$cwa.admin.isEditing" :iri="iri" />
  <!--CWA_MANAGER_END_POSITION-->
</template>

<script setup lang="ts">
import { computed, getCurrentInstance, reactive, ref, toRef } from 'vue'
import { consola } from 'consola'
import ResourceLoader from './ResourceLoader.vue'
import ComponentPlaceholder from './ComponentPlaceholder.vue'
import { useCwa, useCwaResource, useCwaResourceManageable } from '#imports'
import type { IriProp } from '#cwa/runtime/composables/cwa-resource'
import type ManageableComponent from '#cwa/runtime/admin/manageable-component'

const $cwa = useCwa()
const props = defineProps<IriProp>()
const resourceLoader = ref()
const componentManagerOps = reactive({})
const componentResourceManager = ref<undefined|{ manager: ManageableComponent }>()
const iriRef = toRef(props, 'iri')
const resource = useCwaResource(iriRef).getResource()
useCwaResourceManageable(iriRef)

const componentIri = computed(() => {
  const iri = resource.value?.data?.component
  const publishedIri = $cwa.resources.findPublishedComponentIri(iri).value
  if ($cwa.admin.isEditing) {
    const selectedEditingIri = $cwa.admin.componentManager.currentIri.value
    const draftIri = $cwa.resources.findDraftComponentIri(iri).value
    if (selectedEditingIri && [draftIri, publishedIri].includes(selectedEditingIri)) {
      return selectedEditingIri
    }
    return draftIri || iri
  }
  return publishedIri
})

const proxy = getCurrentInstance()?.proxy
if (proxy) {
  componentResourceManager.value = useCwaResourceManageable(componentIri, componentManagerOps, proxy)
} else {
  consola.error('Could not initialise the manager for a component', componentIri.value)
}
const manager = computed(() => componentResourceManager.value?.manager)
</script>
