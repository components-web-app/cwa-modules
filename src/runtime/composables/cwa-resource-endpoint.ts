import type { Ref } from 'vue'
import { computed, onBeforeUnmount, ref, watch, watchEffect } from 'vue'
import { getPublishedResourceState } from '#cwa/runtime/resources/resource-utils'
import { useCwa } from '#cwa/runtime/composables/cwa'

export const useCwaResourceEndpoint = (iri: Ref<string|undefined>, postfix?: string) => {
  const $cwa = useCwa()
  const resource = computed(() => iri.value ? $cwa.resources.getResource(iri.value).value : undefined)
  const forcePublishedVersion = $cwa.admin.resourceStackManager.forcePublishedVersion
  const applyPostfix = ref(false)
  const query = ref('')

  const unwatchEffect = watchEffect(() => {
    if (!resource.value) {
      applyPostfix.value = false
      return
    }
    const publishableState = getPublishedResourceState(resource.value)
    applyPostfix.value = forcePublishedVersion.value !== undefined && publishableState === true
  })

  const unwatchApplyPostfix = watch(applyPostfix, (newApplyPostfix) => {
    if (!newApplyPostfix) {
      query.value = ''
      return
    }
    query.value = forcePublishedVersion.value ? '?published=true' : '?published=false'
  }, {
    immediate: true
  })

  const endpoint = computed(() => `${iri.value}${postfix || ''}${query.value}`)

  onBeforeUnmount(() => {
    unwatchEffect()
    unwatchApplyPostfix()
  })

  return {
    endpoint,
    query
  }
}
