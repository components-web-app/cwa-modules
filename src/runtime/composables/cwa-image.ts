import { computed, type Ref, onMounted, ref } from 'vue'
import { useCwaResourceEndpoint } from '#cwa/runtime/composables/cwa-resource-endpoint'
import { useCwaResource } from '#imports'

export const useCwaImage = (iri: Ref<string>, imagineFilterName?: string) => {
  type MediaFile = {
    contentUrl: string
    fileSize: number
    mimeType: string
    formattedFileSize: string
    imagineFilter?: string
    width?: number
    height?: number
  }

  const cwaResource = useCwaResource(iri)
  const { query } = useCwaResourceEndpoint(iri)
  const resource = cwaResource.getResource()

  const loaded = ref(false)
  const image = ref()

  function handleLoad () {
    loaded.value = true
  }

  const imageFileMediaObjects = computed<MediaFile[]|undefined>(() => {
    return resource.value?.data?._metadata.mediaObjects?.file
  })

  const displayMedia = computed(() => {
    if (!imageFileMediaObjects.value || !imageFileMediaObjects.value.length) {
      return
    }
    const thumbnail = imageFileMediaObjects.value.filter(({ imagineFilter }) => (imagineFilter === imagineFilterName))
    return thumbnail?.[0] || imageFileMediaObjects.value[0]
  })

  const contentUrl = computed(() => {
    const mediaUrl = displayMedia.value?.contentUrl
    if (!mediaUrl) {
      return
    }
    return `${mediaUrl}${query.value}`
  })

  onMounted(() => {
    if (image.value?.complete || image.value?.naturalHeight !== 0) {
      handleLoad()
    }
  })

  return {
    ...cwaResource,
    contentUrl,
    displayMedia,
    handleLoad,
    loaded,
    resource
  }
}
