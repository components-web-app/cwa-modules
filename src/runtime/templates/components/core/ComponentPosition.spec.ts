// @vitest-environment nuxt
import { describe, expect, test, vi } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import { ref } from 'vue'
import * as cwaResourceComposables from '../../../composables/cwaResource'
import ComponentPosition from './ComponentPosition.vue'

const mockComponentName = 'test'

function createWrapper () {
  // @ts-ignore
  vi.spyOn(cwaResourceComposables, 'useCwaResourceUtils').mockImplementation(() => ({
    getResource: vi.fn(() => ref({ data: { component: mockComponentName } }))
  }))

  return shallowMount(ComponentPosition, {
    props: {
      iri: mockComponentName
    }
  })
}

describe('ComponentPosition', () => {
  test('should display ResourceLoader component with componentIri', () => {
    const wrapper = createWrapper()
    const child = wrapper.findComponent({ name: 'ResourceLoader' })
    const { iri, componentPrefix } = child.props()

    expect(iri).toEqual(mockComponentName)
    expect(componentPrefix).toEqual('CwaComponents')
  })

  describe('snapshots', () => {
    test('should match snapshot with ResourceLoader component with componentIri', () => {
      const wrapper = createWrapper()

      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
