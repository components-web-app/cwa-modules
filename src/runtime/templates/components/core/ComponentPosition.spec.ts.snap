// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ComponentPosition > snapshots > should match snapshot with ResourceLoader component with componentIri 1`] = `
<div
  data-v-app=""
>
  
  <!--CWA_MANAGER_START_POSITION-->
  <resourceloader-stub
    componentprefix="CwaComponent"
    iri="test"
  />
  <!--CWA_MANAGER_END_POSITION-->
  
</div>
`;
